xds/src/main/java/io/grpc/xds/internal/sds/DynamicSslContextProvider.java

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright 2020 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds.internal.sds;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import io.envoyproxy.envoy.extensions.transport_sockets.tls.v3.CertificateValidationContext;
+import io.envoyproxy.envoy.extensions.transport_sockets.tls.v3.CommonTlsContext;
+import io.grpc.Status;
+import io.grpc.xds.EnvoyServerProtoData.BaseTlsContext;
+import io.netty.handler.ssl.ApplicationProtocolConfig;
+import io.netty.handler.ssl.SslContext;
+import io.netty.handler.ssl.SslContextBuilder;
+import java.io.IOException;
+import java.security.cert.CertStoreException;
+import java.security.cert.CertificateException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Executor;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+/** Base class for dynamic {@link SslContextProvider}s. */
+public abstract class DynamicSslContextProvider extends SslContextProvider {
+
+  private static final Logger logger = Logger.getLogger(DynamicSslContextProvider.class.getName());
```

Comment:
You can define an abstract `getLogger()`  in `SslContextProvider`, then you will not be using different loggers for different methods.

Reply:
I understand the suggestion but is that pattern being used elsewhere in grpc/xds? With the current pattern any method will always use the logger of the class where the method is defined instead of the logger of the deepest subclass. Since the fully qualified method name in the log output is always that of the method's own class wouldn't it be useful to also have the logger of that class?

Reply:
Good point. Also the current pattern in this PR is the same as what is being used elsewhere.